config/config_handler.go

```diff
@@ -0,0 +1,214 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/configpb"
+	"github.com/pingcap/pd/client"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// ConfHandler is used to load and update config online.
+// See https://github.com/pingcap/tidb/pull/13660 for more details.
+type ConfHandler interface {
+	Start()
+	Close()
+	GetConfig() *Config // read only
+}
+
+// ConfReloadFunc is used to reload the config to make it work.
+type ConfReloadFunc func(oldConf, newConf *Config)
+
+// NewConfHandler creates a new ConfHandler according to the local config.
+func NewConfHandler(localConf *Config, reloadFunc ConfReloadFunc) (ConfHandler, error) {
+	switch defaultConf.Store {
+	case "tikv":
+		return newPDConfHandler(localConf, reloadFunc, nil)
+	default:
+		return &constantConfHandler{localConf}, nil
+	}
+}
+
+type constantConfHandler struct {
+	conf *Config
+}
+
+func (cch *constantConfHandler) Start() {}
+
+func (cch *constantConfHandler) Close() {}
+
+func (cch *constantConfHandler) GetConfig() *Config { return cch.conf }
+
+const (
+	pdConfHandlerRefreshInterval = 30 * time.Second
+	tidbComponentName            = "tidb"
+)
+
+type pdConfHandler struct {
+	id         string // ip:port
+	version    *configpb.Version
+	curConf    atomic.Value
+	interval   time.Duration
+	wg         sync.WaitGroup
+	exit       chan struct{}
+	pdConfCli  pd.ConfigClient
+	reloadFunc func(oldConf, newConf *Config)
+}
+
+func newPDConfHandler(localConf *Config, reloadFunc ConfReloadFunc,
+	newPDCliFunc func([]string, pd.SecurityOption) (pd.ConfigClient, error), // for test
+) (*pdConfHandler, error) {
+	addresses, _, err := ParsePath(localConf.Path)
+	if err != nil {
+		return nil, err
+	}
+	host := localConf.Host
+	if localConf.AdvertiseAddress != "" {
+		host = localConf.AdvertiseAddress
+	}
+	id := fmt.Sprintf("%v:%v", host, localConf.Port)
+	security := localConf.Security
+	if newPDCliFunc == nil {
+		newPDCliFunc = pd.NewConfigClient
+	}
+	pdCli, err := newPDCliFunc(addresses, pd.SecurityOption{
+		CAPath:   security.ClusterSSLCA,
+		CertPath: security.ClusterSSLCert,
+		KeyPath:  security.ClusterSSLKey,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	confContent, err := encodeConfig(localConf)
+	if err != nil {
+		return nil, err
+	}
+
+	// register to PD and get the new default config.
+	// see https://github.com/pingcap/tidb/pull/13660 for more details.
+	// suppose port and security config items cannot be change online.
+	status, version, conf, err := pdCli.Create(context.Background(), new(configpb.Version), tidbComponentName, id, confContent)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if status.Code != configpb.StatusCode_OK {
```

Comment:
`StatusCode_WRONG_VERSION` will return the latest version and config. It should not be returned here?